### PR TITLE
New features to dockerize xknx

### PR DESCRIPTION
**Pull request by mielune**
## Description

To be able to run xknx into a docker container, their is some changes to do:
- automatic local udp port (throw 0) do not allowed us to forward the host's udp port to a named udp port of the container
- using the local ip and port in the Connect, Disconnect and ConnectionState is not working, it is the host ip and port that must be sent to the KNX Gateway.

So new implemented features are:
- Add new optional config `bind_ip` and `bind_port` for connections to be able to dockerised xknx's apps.
- Add optional config `local_port` to be able to map a host udp port to a container udp port.
- Read env vars after reading config file to allow dynamic config for dockerized apps.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] I have performed a self-review of my own code
- [x] The documentation has been adjusted accordingly
- [ ] The changes generate no new warnings
- [ ] Tests have been added that prove the fix is effective or that the feature works
- [x] The changes are documented in the changelog
- [ ] The Homeassistant plugin has been adjusted in case of new config options

### CODE REVIEW

**Comment by farmio** [comment]
Hi!

Thanks for your contribution! 

Have you had a look at #246 ? 
Last time I checked docker was not able to handle multicast connections. If this is still true please remove the configuration additions to routing connections. Also the port for routing connections is configured via xknx.multicast_port, so no need to pass a local_port attribute.

File: xknx/config/config_v1.py [hunk 1]
```diff
@@ -59,6 +62,17 @@ def parse_general(self, doc):
             if "multicast_port" in doc["general"]:
                 self.xknx.multicast_port = doc["general"]["multicast_port"]
 
+    def env_general(self):
```
Comment (marvin-w) [inline xknx/config/config_v1.py:65]: Hint: This is the old config format. We are soon(tm) be migrating to xknx schema v2 - where I will just remove this file in its entirety.

I like the idea of having environment variables in place but I do not agree on the approach taken. I'd prefer having this in the yaml and introduce a dedicated yaml loader for env vars. I can take care of this in the new config schema.

File: xknx_v2.yaml [hunk 2]
```diff
@@ -8,7 +8,10 @@ multicast_group: '224.1.2.3'
 multicast_port: 1337
```
Comment (marvin-w) [inline xknx_v2.yaml:8]: This is the config schema for the new config. You should touch the xknx.yaml if at all.

**Review by marvin-w — CHANGES_REQUESTED** [review]
I like parts of this PR, however I'd have personally implemented them differently. 

I did not yet test the docker part of it. Did you test it? Routing or Tunnel or both? With or without host networking?

**Comment by mielune** [comment]
> Hi! wave
> 
> Thanks for your contribution!
> 
> Have you had a look at #246 ?
> Last time I checked docker was not able to handle multicast connections. If this is still true please remove the configuration additions to routing connections. Also the port for routing connections is configured via xknx.multicast_port, so no need to pass a local_port attribute.

Thanks I didn't see this PR, very closed to mine, I this it is more advance (tests done and some discuss about it).
I will test it in my real env; My code is working fine with tunnel mode.

**Comment by mielune** [comment]
> cked docker was not able to handle

You are right, routing may not work due to multicast. My knx gateway is setup for tunnelling, so I can't test routing mode.
So I will remove this feature from routing mode.

**Comment by mielune** [comment]
This PR is tested in a real env in tunnelling mode, using this features:
- get objects status
- put new objects status
- listen for events from the bus

File: xknx/__version__.py [hunk 3]
```diff
@@ -1,3 +1,3 @@
 """XKNX version."""
 
-__version__ = "0.15.2"
+__version__ = "0.15.3"
```
[the remark this answers is not in the document]
Comment (mielune) [reply xknx/__version__.py]: Done, I rebased from the last master release.

File: xknx/config/config_v1.py (hunk 1 above)
Comment (mielune) [reply xknx/config/config_v1.py:65]: Ok, I will use the new config file.

The practice of loading env after the config file allow us to setup all the stable configuration in the file (devices for ex.), and overwrite all the context dependent attributes from env vars that are set during the container deployment (IPs, ports, etc.). This is a common way of working in container's clusters like Kubernetes and Docker Swarm (I'm using the second one for home automation).

File: xknx_v2.yaml (hunk 2 above)
Comment (mielune) [reply xknx_v2.yaml:8]: v1 condif file updated with all cases of connections.

File: xknx/config/config_v1.py (hunk 1 above)
Comment (marvin-w) [reply xknx/config/config_v1.py:65]: I would still prefer giving the user the ability to change the env vars and also add new ones depending on their needs.

I'm using kubernetes for my home assistant and HA also supports this kind of syntax:

`local_port: !env_var XKNX_LOCAL_PORT`

Something like this. And then you can just use this in your docker compose/deployment.yaml

I prepare a PR for this tomorrow as most of it is already done. I like the other changes though.

**Review by marvin-w — CHANGES_REQUESTED** [review]

File: xknx/config/config_v1.py (hunk 1 above)
Comment (marvin-w) [reply xknx/config/config_v1.py:65]: See https://github.com/XKNX/xknx/pull/480

File: xknx_v2.yaml [hunk 4]
```diff
@@ -8,7 +8,10 @@ multicast_group: '224.1.2.3'
...
 connection: # optional
   type: tunneling # or routing|auto
-  local_ip: "192.168.0.201"
+  local_ip: "192.168.111.201"
+  local_port: 42424
+  bind_ip: "192.168.0.201"
+  bind_port: 42424
```
Comment (marvin-w) [inline xknx_v2.yaml:14]: You'll have to add those to the schema if you want to leave it here.

https://github.com/XKNX/xknx/blob/master/xknx/config/schema.py#L79

Please also add relevant tests for it.

**Review by marvin-w — CHANGES_REQUESTED** [review]
Thank you for the contribution, it looks fine! Please remove all changes (code, changelog entry etc.) related to the env variables as this is covered in another PR and is also handled differently #480.

Please also make sure the build is green and all tests run, otherwise we are not able to merge this.

Happy Hacktoberfest!

**Comment by farmio** [comment]
We (Marvin, Julius and me) did have a conversation back when I did #246 (I think on discord) that the name "proxy_ip" was not ideal and we should pick "masquerade_ip" instead. 
I think "bind_ip" and the attribute name "net_bind" is even more misleading as we are explicitly not binding (as in `socket.bind()`) this address.

**Comment by marvin-w** [comment]
I almost forgot that. It has been a while. But yea, might be worth to change the names of the attributes to something more fitting.

File: docs/configuration.md [hunk 5]
```diff
@@ -23,8 +23,14 @@ The configuration file can contain three sections.
...
     - `gateway_port` (optional) sets the port the KNX tunneling interface is listening on
     - `local_ip` (optional) sets the ip address that is used by xknx
+    - `local_port` (optional) sets the udp port that is used by xknx
+    - `bind_ip` (optional) sets the ip address sent into command telegrams by xknx (for dockerised app)
+    - `bind_port` (optional) sets the udp port sent into command telegrams by xknx (for dockerised app)
   - `routing` for a UDP multicast connection
     - `local_ip` (optional) sets the ip address that is used by xknx
+    - `local_port` (optional) sets the udp port that is used by xknx
+    - `bind_ip` (optional) sets the ip address sent into command telegrams by xknx (for dockerised app)
+    - `bind_port` (optional) sets the udp port sent into command telegrams by xknx (for dockerised app)
```
Comment (farmio) [inline docs/configuration.md:33]:
```suggestion
```
remove these from routing - its tunnelling only.

File: xknx/io/connectionstate.py [hunk 6]
```diff
@@ -7,18 +7,22 @@
 class ConnectionState(RequestResponse):
     """Class to send ConnectonStateRequest and wait for ConnectionStateResponse."""
 
-    def __init__(self, xknx, udp_client, communication_channel_id):
+    def __init__(self, xknx, udp_client, communication_channel_id, net_bind):
```
Comment (farmio) [inline xknx/io/connectionstate.py:10]:
```suggestion
    def __init__(self, xknx, udp_client, communication_channel_id, net_bind=None):
```
make it a keyword attribute like in the other 2 classes

File: xknx/io/request_response.py [hunk 7]
```diff
@@ -16,10 +16,11 @@ class RequestResponse:
 
     # pylint: disable=too-many-instance-attributes
 
-    def __init__(self, xknx, udp_client, awaited_response_class, timeout_in_seconds=1):
+    def __init__(self, xknx, udp_client, awaited_response_class, timeout_in_seconds=1, net_bind=None):
         """Initialize RequstResponse class."""
         self.xknx = xknx
         self.udpclient = udp_client
+        self.net_bind = net_bind
```
Comment (farmio) [inline xknx/io/request_response.py:23]:
```suggestion
    def __init__(self, xknx, udp_client, awaited_response_class, timeout_in_seconds=1):
        """Initialize RequstResponse class."""
        self.xknx = xknx
        self.udpclient = udp_client
```
I'd remove this from the base class and just keep it local in the 3 classes where it is actually used.

File: xknx/io/disconnect.py [hunk 8]
```diff
@@ -7,19 +7,23 @@
...
+        super().__init__(xknx, self.udp_client, DisconnectResponse, net_bind=net_bind)
         self.communication_channel_id = communication_channel_id
 
     def create_knxipframe(self) -> KNXIPFrame:
         """Create KNX/IP Frame object to be sent to device."""
         (local_addr, local_port) = self.udpclient.getsockname()
+        if self.net_bind:
+            endpoint = HPAI(ip_addr=self.net_bind[0], port=self.net_bind[1])
+        else:
+            endpoint = HPAI(ip_addr=local_addr, port=local_port)
```
Comment (farmio) [inline xknx/io/disconnect.py:23]:
```suggestion
        if self.net_bind:
            (local_addr, local_port) = net_bind
        else:
            (local_addr, local_port) = self.udpclient.getsockname()
        endpoint = HPAI(ip_addr=local_addr, port=local_port)
```
we can omit the call to udpclient.getsockname() if we don't use the results anyway. Maybe change the variable names 'local_*' then.

**Review by farmio — CHANGES_REQUESTED** [review]
Some minor changes, apart from renaming `bind_*` and `net_bind`.

**Review by mielune — COMMENTED** [review]
Ok for your reviews, I will make a global review soon.

**Comment by farmio** [comment]
Hi! 
I have just tested setting the IP and port to all `0`. This UDP “Route Back” HPAI works fine with Gira Router and X1. No need to even know the "bind_IP". Maybe you can test this on your system and we can just integrate a `route_back` boolean option in xknx that sets IP and port of outgoing HPAI to 0.

**Comment by mielune** [comment]
> Hi!
> I have just tested setting the IP and port to all `0`. This UDP “Route Back” HPAI works fine with Gira Router and X1. No need to even know the "bind_IP". Maybe you can test this on your system and we can just integrate a `route_back` boolean option in xknx that sets IP and port of outgoing HPAI to 0.

Having free time this holiday, I will test it quickly.

**Comment by mielune** [comment]
> > Hi!
> > I have just tested setting the IP and port to all `0`. This UDP “Route Back” HPAI works fine with Gira Router and X1. No need to even know the "bind_IP". Maybe you can test this on your system and we can just integrate a `route_back` boolean option in xknx that sets IP and port of outgoing HPAI to 0.
> 
> Having free time this holiday, I will test it quickly.

Tried it and working fine.
I add it to a new branch, cleaning the **bind** implementation.

**Comment by farmio** [comment]
Closed if favor for #544

**Pull request closed without merging**